### PR TITLE
[ios] Restore Core Spotlight category search

### PR DESCRIPTION
**Pull request by kirylkaveryn**
Core Spotlight category shortcuts were registered only for the current app
language, and their metadata no longer produced reliable results in iPhone
system search. The index could also remain stale or disappear after iOS purged
it.

This PR restores category results in iOS Spotlight:
- indexes the primary category title for every preferred device language;
- preserves the resolved locale in each identifier, so opening a result uses
  the same localized title and search locale;
- prefers an exact regional locale before the base locale and deduplicates
  titles;
- fully deletes and rebuilds the category domain, so renamed titles do not
  leave stale results;
- rebuilds once per app process, and again after an app version or
  preferred-language change, so a purged system index is restored;
- keeps category shortcuts from expiring between rebuilds;
- persists refresh state only after success and prevents overlapping updates;
- restores the missing Wi-Fi and water thumbnails and verifies all category
  items have images;
- skips real system index updates when the app is launched by tests;
- keeps legacy Spotlight identifiers working and adds focused tests;
- exposes `AppInfo.fullVersion` for reuse by Spotlight and the About screen.

Synonyms are intentionally not indexed yet. iOS Spotlight did not reliably
search `alternateNames` or keywords, while separate synonym entries produced
duplicate and noisy results. Follow-up work is tracked in #13600.

Tested with Xcode 26.6:
- OMaps build-for-testing for iOS Simulator;
- `MWMSearchSpotlightTests`: 17 tests passed;
- iPhone 15 Pro device validation.

Implementation was assisted by Codex.

Screenshot 2026-09-21 at 18 08 19

### CODE REVIEW

File: iphone/Maps/Core/AppDelegate/MapsAppDelegate.mm [hunk 1]
```diff
@@ -97,7 +98,7 @@ - (void)searchText:(NSString *)searchString
     return;
   }
   SearchQuery * query = [[SearchQuery alloc] init:[searchString stringByAppendingString:@" "]
-                                           locale:[MWMSettings spotlightLocaleLanguageId]
+                                           locale:AppInfo.sharedInfo.languageId ?: @"en"
```
Comment (biodranik) [inline iphone/Maps/Core/AppDelegate/MapsAppDelegate.mm]: [P2] Keep the activation query aligned with an indexed locale

This always passes the app primary language while `L(searchStringKey)` localizes into that language. For UI languages absent from `CategoriesHolder` (`az`, `af`, `hy`, `mt`, and others), item generation falls back to another preferred language or English, but activation searches the unsupported UI translation (for example Azerbaijani `Harada yemək` with locale `az`), which cannot match the category data. `SearchTabViewController` already avoids this mismatch. Please carry a supported locale/name in the identifier, or resolve the same effective locale here, and cover an unsupported primary language in the activation tests.

File: iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm [hunk 2]
```diff
@@ -5,67 +5,162 @@
...
+        continue;
+      }
+      [categoryNames addObject:fallbackName];
+    }
 
     NSString * categoryKeyString = @(categoryKey.c_str());
     NSString * imageName = [NSString stringWithFormat:@"ic_%@_spotlight", categoryKeyString];
-    UIImage * image = [UIImage imageNamed:imageName];
-    attrSet.thumbnailData = UIImagePNGRepresentation(image);
+    NSData * thumbnailData = UIImagePNGRepresentation([UIImage imageNamed:imageName]);
```
Comment (biodranik) [inline iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm]: [P3] Make every generated image name resolve

For `category_wifi` and `category_water`, this produces `ic_category_wifi_spotlight` and `ic_category_water_spotlight`, while the compiled asset catalog exports only `ic_wifi_spotlight` and `ic_water_spotlight`. Those Spotlight entries therefore receive a nil thumbnail. Please rename/map the two assets, fail fast in Debug when a displayed-category asset is absent, and assert non-nil thumbnail data for every generated item.

File: iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm [hunk 3]
```diff
@@ -5,67 +5,162 @@
...
++ (void)updateSpotlightIndex:(CSSearchableIndex *)index
+                 languageIds:(NSArray<NSString *> *)languageIds
+                  appVersion:(NSString *)appVersion
+{
+  ASSERT(NSThread.isMainThread, ());
+  // All accesses, including completion handlers, are confined to the main queue.
+  static BOOL indexingInProgress = NO;
+  if (indexingInProgress)
+    return;
+  if ([languageIds isEqualToArray:[MWMSettings spotlightLanguageIds]] &&
```
Comment (biodranik) [inline iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm]: [P2] Recover when Spotlight no longer has the persisted items

These defaults record only that Core Spotlight journaled the last request; they do not prove that the index still contains the items. Core Spotlight can later report index loss or deferred indexing failure through `CSSearchableIndexDelegate`, but no delegate is installed anywhere in the repo. After that, this early return prevents repair until the app version or language list changes. Please implement and acknowledge `reindexAllSearchableItems` / `reindexSearchableItemsWithIdentifiers`, or simplify this to a once-per-process rebuild/upsert instead of persisting the skip across launches.

File: iphone/Maps/Core/AppDelegate/MapsAppDelegate.mm (hunk 1 above)
Comment (kirylkaveryn) [reply iphone/Maps/Core/AppDelegate/MapsAppDelegate.mm]: Fixed in a92164ff29. Spotlight identifiers carry the resolved supported locale, and activation builds the SearchQuery from the same category name and locale. The tests cover an unsupported primary language falling back to the indexed English category.

File: iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm (hunk 2 above)
Comment (kirylkaveryn) [reply iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm]: Fixed in a92164ff29. The Wi-Fi and water asset sets match the generated category image names, generation asserts missing images in Debug, and the tests verify non-empty thumbnail data for every item.

File: iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm (hunk 3 above)
Comment (kirylkaveryn) [reply iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm]: Fixed in a92164ff29 with the suggested once-per-process approach. The first update for each index instance always rebuilds it; successful languages are then tracked in memory, and failed updates remain retryable.

File: iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm [hunk 4]
```diff
@@ -4,68 +4,229 @@
...
+        NSString * currentLocale = localesByCategoryName[categoryName];
+        if (!currentLocale || [locale compare:currentLocale] == NSOrderedAscending)
+          localesByCategoryName[categoryName] = locale;
+        break;
+      }
+    }
+    if (categoryNames.count == 0)
+    {
+      NSString * fallbackName = localizedCategoryNames[@"en"];
+      ASSERT(fallbackName.length > 0, (categoryKey));
```
Comment (biodranik) [inline iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm]: [P3] Fail fast on missing bundled category data

All 23 `DisplayedCategories` keys have an English primary name; absence is a bundled-data/programming error, not recoverable input. In Release this branch silently omits the category and then persists a successful partial rebuild. Please use `CHECK(fallbackName.length > 0, ...)` and remove the log/continue fallback; apply the same invariant to the activation fallback below.

File: iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm [hunk 5]
```diff
@@ -4,68 +4,229 @@
...
+    locale = @"en";
+    categoryName = GetCategoryName(categoryKey.UTF8String, locale);
   }
+  ASSERT(categoryName.length > 0, (categoryKey.UTF8String));
+  if (categoryName.length == 0)
+    return nil;
+
+  return [[SearchQuery alloc] init:[categoryName stringByAppendingString:@" "]
+                            locale:locale
+                            source:SearchTextSourceDeeplink];
```
Comment (biodranik) [inline iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm]: [P2] Mark Spotlight shortcuts as category searches

These items come exclusively from `DisplayedCategories`, but `Deeplink` leaves `MWMSearch::m_isCategory` false. The core then also matches the title as ordinary name/address tokens and uses normal ranking, while `SearchTabViewController` passes `SearchTextSourceCategory` for the same shortcuts to request pure category results. Please use `SearchTextSourceCategory` here and assert the source in the activation tests.

File: iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm [hunk 6]
```diff
@@ -4,68 +4,229 @@
...
+  ASSERT(NSThread.isMainThread, ());
+  // All accesses, including completion handlers, are confined to the main queue.
+  static BOOL indexingInProgress = NO;
+  static NSHashTable<CSSearchableIndex *> * refreshedIndexes =
+      [NSHashTable hashTableWithOptions:NSPointerFunctionsWeakMemory | NSPointerFunctionsObjectPointerPersonality];
+  if (indexingInProgress)
+    return;
+  // Rebuild once per process so a purged system index is repaired on the next app launch.
+  if ([refreshedIndexes containsObject:index] && [languageIds isEqualToArray:[MWMSettings spotlightLanguageIds]] &&
+      [appVersion isEqualToString:[MWMSettings spotlightAppVersion]])
```
Comment (biodranik) [inline iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm]: [P3] Remove persisted state that cannot affect production anymore

Because `refreshedIndexes` starts empty, the first call in every process always rebuilds. The app version cannot change later in that process, so `spotlightAppVersion` never drives another production rebuild; the next installed version starts a new process and already rebuilds. Persisted languages are likewise only consulted after an in-process success. A weak-key `NSMapTable<CSSearchableIndex *, NSArray<NSString *> *>` holding the last successful languages per index can replace `refreshedIndexes` plus both UserDefaults values, remove the `appVersion` argument/settings, and keep injected indexes independent without changing behavior.

File: iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm (hunk 4 above)
Comment (kirylkaveryn) [reply iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm]: Fixed in a92164ff29. Both bundled-data fallbacks use CHECK, so a missing English primary category name cannot produce a successful partial index.

File: iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm (hunk 5 above)
Comment (kirylkaveryn) [reply iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm]: Fixed in a92164ff29. Spotlight activation creates SearchTextSourceCategory queries, and the activation test asserts the source for every localized item.

File: iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm (hunk 6 above)
Comment (kirylkaveryn) [reply iphone/Maps/Core/Search/MWMSearch+CoreSpotlight.mm]: Fixed in a92164ff29. A weak-key NSMapTable stores the last successful language list per index. The appVersion argument and Spotlight UserDefaults state were removed, together with the no-longer-needed AppInfo.fullVersion refactor.

**Review by biodranik — APPROVED** [review]
Thanks, LGTM! How to announce this feature for iOS users in release notes? And how to test it properly?
